comicapi/genericmetadata.py

```diff
@@ -202,24 +252,129 @@ def get_clean_metadata(self, *attributes: str) -> GenericMetadata:
+        """Overwrite (no clear/reset) - Same as Overwrite but None/empty string from 'new' do NOT overwrite values in 'self'."""
+        if new is not None:
+            if isinstance(new, (list, set)) and len(new) > 0 and isinstance(getattr(self, cur), (list, set)):
+                setattr(self, cur, new)
+            elif isinstance(new, str) and len(new) > 0:
+                setattr(self, cur, new)
+            elif isinstance(new, (int, float, bool)):
+                setattr(self, cur, new)
+
+    def assign_add_missing(self, cur: str, new: Any) -> None:
+        """Add Missing - Any values that are None or an empty str/list/set, add them from 'new' to 'self'."""
+        if new is not None:
+            if (
+                isinstance(getattr(self, cur), (list, set))
+                and len(getattr(self, cur)) == 0
+                and isinstance(new, (list, set))
+                and len(new) > 0
+            ):
+                setattr(self, cur, new)
+            elif (
+                isinstance(getattr(self, cur), str)
+                and len(getattr(self, cur)) == 0
+                and isinstance(new, str)
+                and len(new) > 0
+            ):
+                setattr(self, cur, new)
+            elif getattr(self, cur) is None:
+                setattr(self, cur, new)
+
+    def assign_combine_new(self, cur: str, new: Any) -> None:
+        """Combine (New) - Combine lists and sets. Replace all None values. Any other values that collide the 'new' value is used (ignoring None and empty strings)."""
+        if new is not None:
+            if isinstance(new, (list, set)) and isinstance(getattr(self, cur), (list, set)):
+                self.assign_dedupe(new, cur, True)
+            elif (
+                isinstance(new, str)
+                and len(new) > 0
+                and (isinstance(getattr(self, cur), str) or getattr(self, cur) is None)
+            ):
+                setattr(self, cur, new)
+            else:
+                # int, float and bool
+                setattr(self, cur, new)
+
+    def assign_combine_cur(self, cur: str, new: Any) -> None:
+        """Combine (Current) - Same as Combine (New) but 'self' value is kept with collisions."""
+        if new is not None:
+            if isinstance(new, (list, set)) and isinstance(getattr(self, cur), (list, set)):
+                self.assign_dedupe(new, cur)
+            elif isinstance(new, str) and len(getattr(self, cur)) == 0:
+                setattr(self, cur, new)
+            elif getattr(self, cur) is None:
+                setattr(self, cur, new)
+
+    def overlay(self, new_md: GenericMetadata, mode: OverlayMode = OverlayMode.overlay) -> None:
+        """Overlay a new metadata object on this one"""
+
+        assign_funcs = {}
+        for mem in OverlayMode:
+            assign_funcs[mem.name] = getattr(self, "assign_" + mem.name)
 
         def assign(cur: str, new: Any) -> None:
```

Comment:
There's a whole lot of processing going on here that doesn't need to happen. This dict can just be manually created.

Instead of doing any of that how about this
```python
def assign_overlay(md: GenericMetadata, attribute: str, new_value: Any) -> None:
    """Overlay - When the new object has non-None values, over-write them to this one."""
    if new_value is not None:
        if isinstance(new_value, str) and len(new_value) == 0:
            if isinstance(getattr(md, attribute), (list, set)):
                getattr(md, attribute).clear()
            else:
                setattr(md, attribute, None)
        elif isinstance(new_value, (list, set)) and len(new_value) == 0:
            pass
        else:
            setattr(md, attribute, new_value)
...
class GenericMetadata:
    ...
    def overlay(self, new_md: GenericMetadata, assign=assign_overlay) -> None:
        """Overlay a new metadata object on this one"""

        if not new_md.is_empty:
            self.is_empty = False

        assign(self, "tag_origin", new_md.tag_origin)
```
Or instead of keeping the current function signature
```python
    def overlay(self, new_md: GenericMetadata, assign=assign_overlay) -> None:
        """Overlay a new metadata object on this one"""

        if not new_md.is_empty:
            self.is_empty = False

        self.tag_origin = assign(self.tag_origin, new_md.tag_origin)
```

Reply:
This (`self.tag_origin = assign(self.tag_origin, new_md.tag_origin)`) now makes assign not representative of what it is doing so should be renamed

Reply:
``` python
    def overlay(self, new_md: GenericMetadata, assign=assign_overlay) -> None:
        """Overlay a new metadata object on this one"""

        if not new_md.is_empty:
            self.is_empty = False

        self.tag_origin = assign(self.tag_origin, new_md.tag_origin)
```
The `overlay` gets called from outside and will pass in the `config[0].Comic_Book_Lover__metadata_overlay` for the `assign` so keeping `def overlay(self, new_md: GenericMetadata, mode: OverlayMode = OverlayMode.overlay) -> None:` makes the most sense? Along with:
```python
assign_funcs = {
            "overlay": self.assign_overlay,
            "overwrite": self.assign_overwrite,
            "overwrite_no_clear": self.assign_overwrite_no_clear,
            "add_missing": self.assign_add_missing,
            "combine_new": self.assign_combine_new,
            "combine_current": self.assign_combine_current,
        }

        def assign(cur: str, new: Any) -> None:
            assign_func = assign_funcs.get(mode.name, self.assign_overlay)
            assign_func(cur, new)
```

`self.tag_origin = assign(self.tag_origin, new_md.tag_origin)`
I did think having a return would be easier now there is multiple options, so I will change it to that.

Reply:
> This (`self.tag_origin = assign(self.tag_origin, new_md.tag_origin)`) now makes assign not representative of what it is doing so should be renamed if used

Should `tag_origin` be changed to a `set[tag_origin]`?

Reply:
> Should `tag_origin` be changed to a `set[tag_origin]`?

Let's leave it a list for now



> ```python
> assign_funcs = {
>             "overlay": self.assign_overlay,
>             "overwrite": self.assign_overwrite,
>             "overwrite_no_clear": self.assign_overwrite_no_clear,
>             "add_missing": self.assign_add_missing,
>             "combine_new": self.assign_combine_new,
>             "combine_current": self.assign_combine_current,
>         }
> ```

You're so close now replace these strings with the enums
`"overlay" -> OverlayMode.overlay`

Reply:
And then the assign function you can change to
```python
assign = assign_funcs.get(mode, self.assign_overlay)
```

Reply:
Adding a date doesn't make sense as there is nothing that would actually set it in a useable way. CT would set it and then it would save it and then it would disappear.